enhancements/microshift/kubernetes-for-device-edge.md

```diff
@@ -0,0 +1,508 @@
+for configuring the operating system components and anything that runs
+on the device outside of MicroShift.
+
+Administrators of edge devices may have little or no control over the
+network to which the devices are attached, especially the DNS or DHCP
+services on those networks. Edge devices are unlikely to have DNS
+entries added for MicroShift. MicroShift itself must also tolerate the
+host changing IP address when a DHCP lease expires or the host
+reboots.
+
+Many of the OpenShift APIs are not relevant for edge device use cases.
+We are starting with the absolute minimal set for now because it will
+be easier to add more APIs later than it will be to remove anything.
+It is meant as a pure runtime tool, and will not be used for
+build/dev-cluster use cases, therefore we will not include Builds,
+BuildConfigs, etc.  MicroShift is not an interactive, multi-user
+environment, therefore we do not include some of the authorization
+APIs such as users, projects, etc.  It doesn't manage OS or
+infrastructure, so we do not include MachineConfig and related APIs.
+MicroShift is a monolith, therefore we do not include any cluster
+configuration APIs.
+
+In addition to the standard APIs included in core Kubernetes, the
+OpenShift APIs supported will be:
+
+* route.openshift.io/v1 for Routes
+* security.openshift.io/v1 for SecurityContextConstraints
+* authorization.openshift.io/v1 for RBAC APIs such as Role,
+  RoleBinding, etc.
+
+In addition to those APIs, we will enable two OpenShift controllers
+related to Ingress:
+
+* openshift.io/ingress-ip
+* openshift.io/ingress-to-route
+
+MicroShift is built downstream of OpenShift to ensure that it includes
+the same versions of the embedded components that are present in
+standalone OpenShift, allowing us to maintain binary compatibility.
+The versions of the embedded components are selected by looking at an
+OpenShift release image, finding the git SHAs for the direct
+dependencies, resolving the secondary dependencies, and vendoring the
+results into the MicroShift git repository. That process needs to be
+as automatic as possible, so we want to avoid carrying any
+MicroShift-specific patches that must be applied after a rebase. The
+version of the containerized components are taken from the same
+release image and the image references are built into the MicroShift
+binary.
+
+A great deal of effort has gone into making OpenShift 4 self-managing,
+especially when it comes to multi-node deployment and upgrades. We do
+not want to reinvent that work, nor do we want to recreate the
+OpenShift 3 model using an outside orchestration tool.  A single-node
+control plane becomes a single point of failure when worker nodes are
+attached, which eliminates the high availability benefits of running
+workloads on Kubernetes. Those statements taken together lead us to
+conclude that MicroShift will only support deployments with 1 node
+acting as both control plane and worker.
```

Comment:
Those statements apply to SNO as well, but there was a recent decision to allow for the addition of worker nodes.  This makes sense for use cases requiring higher computing capacity, but having a SPOF is acceptable.  Having multiple small edge devices forming a cluster doesn't match the use cases as we understand them today, and it makes sense to not allow multi-node deployments at this time.  However, I don't think we can set in stone that it will never happen, as the industry will continue to evolve and new use cases will arise.